core/src/sumeragi/mod.rs

```diff
@@ -205,33 +205,32 @@ impl SumeragiHandle {
             kura,
             network,
             genesis_network,
-            block_hashes,
+            block_count,
         }: SumeragiStartArgs,
     ) -> SumeragiHandle {
         let (message_sender, message_receiver) = mpsc::sync_channel(100);
 
-        for (block_hash, i) in block_hashes
-            .iter()
-            .take(block_hashes.len().saturating_sub(1))
-            .zip(1u64..)
-        {
-            let block_height: u64 = i;
-            let block_ref = kura.get_block_by_height(block_height).expect("Sumeragi could not load block that was reported as present. Please check that the block storage was not disconnected.");
-            assert_eq!(
-                block_ref.hash(),
-                *block_hash,
-                "Kura init correctly reported the block hash."
-            );
-
-            wsv.apply(&block_ref)
+        let mut blocks_iter = (1..=block_count).map(|block_height| {
+            kura.get_block_by_height(block_height as u64).expect("Sumeragi could not load block that was reported as present. Please check that the block storage was not disconnected.")
+        });
+
+        let block_iter_except_last = (&mut blocks_iter).take(block_count.saturating_sub(1));
+        for block in block_iter_except_last {
+            block
+                .revalidate(&mut wsv)
+                .expect("Failed to revalidate block in init.");
+            wsv.apply_without_execution(block.as_ref())
                 .expect("Failed to apply block to wsv in init.");
         }
-        let finalized_wsv = wsv.clone();
 
-        if !block_hashes.is_empty() {
-            let block_ref = kura.get_block_by_height(block_hashes.len() as u64).expect("Sumeragi could not load block that was reported as present. Please check that the block storage was not disconnected.");
+        // finalized_wsv is one block behind
+        let finalized_wsv = wsv.clone();
 
-            wsv.apply(&block_ref)
+        if let Some(latest_block) = blocks_iter.next() {
+            latest_block
+                .revalidate(&mut wsv)
+                .expect("Failed to revalidate block in init.");
```

Comment:
```suggestion
                .expect("The block should be valid in init");
```

Or.. what assumption is expected? Why "init" assumes that revalidation should never fail?

Reply:
> Why "init" assumes that revalidation should never fail?
 
Because we store only valid blocks in the block store.

Reply:
Can we trust that block store cannot be tampered with?

Reply:
We can't but what best course of action would be in such case?

We panic so user have to manually inspect what happen in the block store (e.g. clear it).

Reply:
> Can we trust that block store cannot be tampered with?

Actually this is an involved question. 

1. It may be subject to bit-rot. We don't have a mechanism for counteracting this, but it's trivial to fix and detect: check if any of the blocks in the block-store has the wrong hash in an `async` task, and if so, refresh from other peers. 

2. The block could be tampered with maliciously. This could be either because a party gained access to the peer via a CVE, or because the peer administrator was corrupt. At this point all we need to do is convey to the peer admin that if they continue using an invalid block store that they'd lose money. If you can tamper with the block store, you can easily tamper with a build of a ledger to which you have the source code. 

### Corollary

The original message of "Failed to revalidate block at initialisation" is adequate.

I _would_, if possible, tell the user how to fix the block store, but given that it's now a single file with multiple blocks in binary, we'd need extra tooling or very good docs.